crates/subspace-farmer/src/farmer_cache/piece_cache_state.rs

```diff
@@ -0,0 +1,150 @@
+                    })
+                    .collect::<Vec<_>>();
+                sorted_backends.sort_unstable_by_key(|(_, backend)| backend.free_size());
+                sorted_backends
+                    .into_iter()
+                    .rev()
+                    .find_map(|(cache_index, backend)| {
+                        backend
+                            .next_free()
+                            .map(|free_offset| FarmerCacheOffset::new(cache_index, free_offset))
+                    })
+            }
+        }
+    }
+
+    pub(super) fn get_stored_piece(
+        &self,
+        key: &RecordKey,
+    ) -> Option<&FarmerCacheOffset<CacheIndex>> {
+        self.stored_pieces.get(key)
+    }
+
+    pub(super) fn contains_stored_piece(&self, key: &RecordKey) -> bool {
+        self.stored_pieces.contains_key(key)
+    }
+
+    pub(super) fn push_stored_piece(
+        &mut self,
+        key: RecordKey,
+        cache_offset: FarmerCacheOffset<CacheIndex>,
+    ) -> Option<FarmerCacheOffset<CacheIndex>> {
+        self.stored_pieces.insert(key, cache_offset)
+    }
+
+    pub(super) fn stored_pieces(&self) -> Values<'_, RecordKey, FarmerCacheOffset<CacheIndex>> {
+        self.stored_pieces.values()
+    }
+
+    pub(super) fn remove_stored_piece(
+        &mut self,
+        key: &RecordKey,
+    ) -> Option<FarmerCacheOffset<CacheIndex>> {
+        self.stored_pieces.remove(key)
+    }
+
+    pub(super) fn free_stored_piece_if<F>(&mut self, pred: F)
+    where
+        F: FnMut(&RecordKey, &mut FarmerCacheOffset<CacheIndex>) -> bool,
+    {
+        self.stored_pieces
+            .extract_if(pred)
+            .for_each(|(_piece_index, offset)| {
+                // There is no need to adjust the `last_stored_offset` of the `backend` here,
+                // as the free_offset will be preferentially taken from the dangling free offsets
+                self.dangling_free_offsets.push_back(offset);
+            })
+    }
+
+    pub(super) fn push_dangling_free_offset(&mut self, offset: FarmerCacheOffset<CacheIndex>) {
+        trace!(?offset, "Pushing dangling free offset");
+        self.dangling_free_offsets.push_back(offset);
+    }
+
+    pub(super) fn get_backend(&self, cache_index: CacheIndex) -> Option<&CacheBackend> {
+        self.backends.get(usize::from(cache_index))
+    }
+
+    pub(super) fn backends(&self) -> impl Iterator<Item = &CacheBackend> {
+        self.backends.iter()
+    }
+
+    pub(super) fn backends_count(&self) -> usize {
+        self.backends.len()
+    }
+
+    pub(super) fn clear(
+        self,
+    ) -> (
+        HashMap<RecordKey, FarmerCacheOffset<CacheIndex>>,
+        VecDeque<FarmerCacheOffset<CacheIndex>>,
+    ) {
+        let Self {
+            mut stored_pieces,
+            mut dangling_free_offsets,
+            backends: _,
+        } = self;
+
+        stored_pieces.clear();
+        dangling_free_offsets.clear();
+        (stored_pieces, dangling_free_offsets)
+    }
```

Comment:
I really don't like this stuff. `clear` method usually takes `&mut self` and returns nothing. You don't necessarily need to literally create a method for every single call. When you do create a method, it is useful to pick a sensible name. Names like `.push()` and `.pop()` are too generic. Here a better name would be `fn reuse()` or `fn repurpose()` because the goal here is to reuse existing memory allocations, not to clear something.